src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -32,145 +26,203 @@
-				"[the] line %number% [of %block%]", "[the] (1¦1st|1¦first|2¦2nd|2¦second|3¦3rd|3¦third|4¦4th|4¦fourth) line [of %block%]");
+			"[the] lines [of %block%]",
+			"[the] line %number% [of %block%]",
+			"[the] (1¦1st|1¦first|2¦2nd|2¦second|3¦3rd|3¦third|4¦4th|4¦fourth) line [of %block%]"
+			);
 	}
-	
+
 	private static final ItemType sign = Aliases.javaItemType("sign");
-	
+
+	private boolean isLines;
 	@SuppressWarnings("null")
 	private Expression<Number> line;
 	@SuppressWarnings("null")
-	private Expression<Block> block;
-	
-	@SuppressWarnings({"unchecked", "null"})
-	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		if (matchedPattern == 0)
-			line = (Expression<Number>) exprs[0];
-		else
-			line = new SimpleLiteral<>(parseResult.mark, false);
-		block = (Expression<Block>) exprs[exprs.length - 1];
-		return true;
-	}
-	
+	private Expression<Block> exprBlock;
+	private boolean delayed;
+
 	@Override
-	public boolean isSingle() {
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		isLines = (matchedPattern == 0);
+		if (!isLines)
+			line = matchedPattern == 1 ? (Expression<Number>) exprs[0] : new SimpleLiteral<>(parseResult.mark, false);
+		exprBlock = (Expression<Block>) exprs[exprs.length - 1];
+		delayed = isDelayed == Kleenean.TRUE;
 		return true;
 	}
-	
-	@Override
-	public Class<? extends String> getReturnType() {
-		return String.class;
-	}
-	
+
 	@Override
 	@Nullable
-	protected String[] get(final Event e) {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return new String[0];
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return new String[0];
-		if (getTime() >= 0 && block.isDefault() && e instanceof SignChangeEvent && !Delay.isDelayed(e)) {
-			return new String[] {((SignChangeEvent) e).getLine(line)};
+	protected String[] get(Event event) {
+		Block block = exprBlock.getSingle(event);
+		if (isLines) {
+			if (block == null || !sign.isOfType(block))
+				return new String[0];
+			if ((event instanceof SignChangeEvent && getTime() >= 0 && block.equals(((SignChangeEvent) event).getBlock()) && !delayed))
+				return ((SignChangeEvent) event).getLines();
+			return ((Sign) (block.getState())).getLines();
+		} else {
+			Number lineNumber = line.getSingle(event);
+			if (lineNumber == null || lineNumber.intValue() < 1 || lineNumber.intValue() > 4)
+				return new String[0];
+			int index = lineNumber.intValue() - 1;
+			if (event instanceof SignChangeEvent && getTime() >= 0 && block == ((SignChangeEvent) event).getBlock() && !delayed) {
+				return CollectionUtils.array(((SignChangeEvent) event).getLine(index));
+			}
+			if (block == null || !sign.isOfType(block)) {
+				return new String[0];
+			}
+			return CollectionUtils.array(
+				((Sign) (block.getState())).getLine(index)
+			);
```

Comment:
This is already short enough to where it doesn't need multi line breaks.
```suggestion
			return CollectionUtils.array(((Sign) (block.getState())).getLine(index));
```

Reply:
Oh, I broke it down as I thought of the amount of paranthesis will cause reading issues, but sure thing if it's fine having it in one line